src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala

```diff
@@ -0,0 +1,574 @@
+
+    val readers: Seq[SequentialFastqSource] = (fastq +: fastqs.toSeq).map { otherFastqs: Seq[Path] =>
+      if (otherFastqs.size != fastq.size) throw new IllegalArgumentException("List of list fastqs had differing lengths.")
+      SequentialFastqSource(otherFastqs)
+    }
+
+    def next: Seq[FastqRecord] = {
+      val records = readers.map(_.next())
+      verifySameReadNames(records)
+      records
+    }
+
+    def hasNext: Boolean = {
+      if (readersHaveNext(true)) true
+      else if (readersHaveNext(false)) false
+      else throw new IllegalArgumentException("Fastqs have differing lengths")
+    }
+
+    private def readersHaveNext(desiredValue: Boolean): Boolean = readers.forall(_.hasNext == desiredValue)
+
+    private def verifySameReadNames(records: Seq[FastqRecord]): Unit = {
+      val readName = records.head.name
+      if (records.exists(_.name != readName)) {
+        throw new IllegalArgumentException(String.format("Mismatching read names in FASTQS:\n%s\n%s\n", readName, records.find(_.name != readName).get))
+      }
+    }
+
+    def close() = readers.foreach(_.close())
+  }
+
+  /** Helper class to convert fastq records into SAM records and write them out.
+    *
+    * @param sampleSheet the sample sheet describing all the samples
+    * @param readStructures the read structures info, one for each expected fastq record that will be passed
+    *                           into `convertFastqsAndWriteRecords`.
+    */
+  private class FastqConverter(val sampleSheet: SampleSheet,
+                               val umiTag: String,
+                               val multiUmiTags: Seq[String],
+                               val qualityFormat: FastqQualityFormat,
+                               val minQ: Int,
+                               val maxQ: Int,
+                               val maxMismatches: Int,
+                               val minMismatchDelta: Int,
+                               val maxNoCalls: Int,
+                               val readStructures: ReadStructure*) {
+    /** Sample barcode to its associate metric */
+    private val barcodeToMetrics: mutable.Map[String, SampleBarcodeMetric] = new mutable.LinkedHashMap[String, SampleBarcodeMetric]()
+
+    sampleSheet.foreach { sample =>
+      val barcode: String = sample.sampleBarcode match {
+        case Some(b) => b.concatenatedBarcode
+        case None => throw new IllegalArgumentException(s"Sample with id '${sample.sampleId}' did not have a sampleBarcode")
+      }
+      val libraryId: String = sample.libraryId match {
+        case Some(id) => id
+        case None => throw new IllegalArgumentException(s"Sample with id '${sample.sampleId}' did not have a library id")
+      }
+      val metric: SampleBarcodeMetric = SampleBarcodeMetric(sample.sampleName, libraryId, barcode)
+      barcodeToMetrics.put(barcode, metric)
+    }
+
+    private val noMatchBarcode = readStructures.flatMap {
```

Comment:
should not be stored as a member, but instead created once and looked up in `barcodeToMetrics`